nomad/eval_endpoint.go

```diff
@@ -230,7 +230,8 @@ func (e *Eval) Ack(args *structs.EvalAckRequest,
 	if err := e.srv.evalBroker.Ack(args.EvalID, args.Token); err != nil {
 		return err
 	}
-	return nil
+
+	return cancelCancelableEvals(e.srv)
```

Comment:
Hm, so this introduces a raft operation in `Ack(...)` which didn't have one before.

Since cancelable evals are already moved out of the blocked queue, we don't *need* to commit them to make the optimization work. However the longer we wait between updating the Ack'd eval and committing canceled evals, the longer the statestore will confusingly have `ready` evals that aren't be processed because they're sitting in the cancelable queue... :thinking: 

Since your benchmarking demonstrated this approach is already a massive improvement, let's leave this be.

However, I think we could make this async (maybe by just making it wakeup the reapCancelableEvaluations loop?) and potentially speed things up further.


If I'm right it might be worth commenting because it will be very hard for futures folks to load the context to make this determination:

```suggestion
	// It's not necessary to cancel evals before Ack returns, but it's done here to
	// commit canceled evals as close to the Ack'd eval being committed as possible.
	return cancelCancelableEvals(e.srv)
```

Reply:
Actually, the wake-up channel is probably a safer implementation -- if it's a buffered channel that'll ensure that `Ack` never gets blocked on raft. I've done that here. That also lets me move the `cancelCancelableEvals` into `leader.go`, as it now has only that one caller.

Reply:
Did you mean to change this call to use the wakeup chan, because it's still performing the raft apply synchronously on main.

I don't think it's a big deal, especially because your benchmark demonstrated it's an immense speedup already.

Reply:
🤦  Yes. Thanks for catching that. This line should now be 

```go
s.reapCancelableEvalsCh <-struct{}
return nil
```

Reply:
Fixed in https://github.com/hashicorp/nomad/pull/15294